Issue #1367: Window parallelism

hawkfish opened the pull request:
Implement multi-threaded, hashed, partitioned window operator.

Commit:
Issue #1367: Window hash partition

Use hash partitioning to break an OVER clause up
into 1024 disjoint partitions. Sort these partitions
independently.

The code is currently single-threaded, but there are now
many opportunities for parallelisation.

Commit:
Issue #1367: Window hash partition

Fix formatting.

Commit:
Issue #1367: Window hash partition

Another non-deterministic test.

Commit:
Issue #1367: Window hash partition

More non-deterministic tests.

Commit:
Issue #1367: Window partitions

Refactor utility functions to take row
Slices instead of working on all partitions
at once.

Commit:
Issue #1367: Window partitions

Compute window functions on partitions instead of
on the entire table.

Commit:
Merge branch 'hawkfish-window-partition' of github.com:hawkfish/duckdb into hawkfish-window-partition

Commit:
Issue #1367: Window hash partition

Fix fencepost error in segment tree creation.
Offset segment tree ranges for call.
Suppress linting for canonical container inner class name.

Commit:
Issue #1367: Window hash partitioning

Use larger ranges for computing window functions.
Pre-populate the partition bitmap with all
the boundaries to avoid redundant Value comparisons.

Commit:
Issue #1367: Window parallelisation

Compute the window functions
with a shared OVER clause in parallel
broken downby chunk-sized partition
sets.

Commit:
Issue 1367: Window parallelism

Fix tidy issues.

hannesmuehleisen commented:
So we are currently re-implementing the Physical Sort operator to be out-of-core capable and parallel, I wonder whether it might be better to use this operator instead of the chunk collection sorter in the window operator. This could be done in the planner...

hawkfish commented:
I did't parallelise the partition sorting yet (or the partitioning itself) but I think these could be orthogonal. The bulk of this CL (aside from the bug fix in the segment tree creator) was directed at parallel function evaluation and hash partitioning because I knew sorting was being attacked separately. The hash code in the CL does not use the `Value` interfaces and is set up to sort partitions instead of the entire table. That said, if there is only one partition, sorting becomes the bottleneck.

As for planning, that would be along the lines of Cao et al.? I was thinking that determining the optimal key pair ordering was something that could be done inside the `PhysicalWindow`.

hawkfish commented:
So is the issue that `SortPartitionedCollectionForWindow` makes use of `SortSlice` and `ReorderSlice`?  I'd be happy to help with the transition to the new sorting paradigm.

hannesmuehleisen commented:
It's not so much an issue, more an attempt at clarifying that improvements of chunk collection sorting are merely temporary. I think the transition is also a different PR once the new sorting has been merged.

hannesmuehleisen commented:
CC @mbasmanova

hannesmuehleisen commented:
Found an issue with a R-based test case. `Type DOUBLE does not match with INTEGER. Type mismatch when combining rows`. Will try to port to a normal test.

hawkfish commented:
Is this in my CL or on `master`?

hannesmuehleisen commented:
On this PR. Test case is coming.

hannesmuehleisen commented:
Here we go:

````
statement ok
CREATE  TABLE "data" ("Store" INTEGER, "Dept" INTEGER, "Date" DATE, "Weekly_Sales" DOUBLE, "IsHoliday" BOOLEAN);

statement ok
insert into data select * from read_csv_auto('test/walmart_train.csv');

statement ok
pragma threads=8;

statement ok
  SELECT
  *,
  lag(Weekly_Sales, 1) OVER(PARTITION BY "Store", "Dept" ORDER BY "Date" ASC) AS Lag_1,
  lag(Weekly_Sales, 2) OVER(PARTITION BY "Store", "Dept" ORDER BY "Date" ASC) AS Lag_2,
  lag(Weekly_Sales, 3) OVER(PARTITION BY "Store", "Dept" ORDER BY "Date" ASC) AS Lag_3,
  lag(Weekly_Sales, 4) OVER(PARTITION BY "Store", "Dept" ORDER BY "Date" ASC) AS Lag_4,
  lag(Weekly_Sales, 5) OVER(PARTITION BY "Store", "Dept" ORDER BY "Date" ASC) AS Lag_5,
  lag(Weekly_Sales, 6) OVER(PARTITION BY "Store", "Dept" ORDER BY "Date" ASC) AS Lag_6,
  lag(Weekly_Sales, 7) OVER(PARTITION BY "Store", "Dept" ORDER BY "Date" ASC) AS Lag_7,
  lag(Weekly_Sales, 8) OVER(PARTITION BY "Store", "Dept" ORDER BY "Date" ASC) AS Lag_8,
  lag(Weekly_Sales, 9) OVER(PARTITION BY "Store", "Dept" ORDER BY "Date" ASC) AS Lag_9,
  lag(Weekly_Sales, 10) OVER(PARTITION BY "Store", "Dept" ORDER BY "Date" ASC) AS Lag_10
FROM
  data;
````

This fails for me with `pointer being freed was not allocated`

Dataset is here: https://www.dropbox.com/s/2str3ek4f4cheqi/walmart_train.csv?dl=1

Mytherin commented on test/sql/window/test_window_dbplyr.test:
```diff
@@ -15,15 +15,15 @@ INSERT INTO dbplyr_052 VALUES (1,1, 42),(2,1, 42),(3,1, 42),(2,2, 42),(3,2, 42),
 # this breaks because we add a fake projection that is not pruned
 query IR
-SELECT x, g FROM (SELECT x, g, SUM(x) OVER (PARTITION BY g ORDER BY x ROWS UNBOUNDED PRECEDING) AS zzz67 FROM (SELECT x, g FROM dbplyr_052 ORDER BY w) dbplyr_053) dbplyr_054 WHERE (zzz67 > 3.0)
+SELECT x, g FROM (SELECT x, g, SUM(x) OVER (PARTITION BY g ORDER BY x ROWS UNBOUNDED PRECEDING) AS zzz67 FROM (SELECT x, g FROM dbplyr_052 ORDER BY w) dbplyr_053) dbplyr_054 WHERE (zzz67 > 3.0) ORDER BY x, g
```
Are there any tests for the parallel window functions? It would be great if you could add some under `test/sql/parallelism/intraquery`. The SQL server tests (`test\sqlserver\test_sqlserver.test_slow`) would be a good starting point as they contain many window functions. You can add the following lines on top there to force the parallel window functions to be tested:

```sql
statement ok
PRAGMA threads=4

statement ok
PRAGMA force_parallelism
```

It would be great to see some more simple/stand-alone tests as well.

hawkfish replied:
I can do that, but to some extent I was assuming that my changes were a NOP. The test that @hannesmuehleisen found above seems to be a good example (although I think it may be too big?)

Mytherin replied:
On the contrary I would think introducing parallelism to an operator is an easy way to introduce many subtle bugs :) I would prefer to see this tested well.

hawkfish replied:
`test/sql/window/test_over_grouping.test_slow` was added in a previous CL for #1367 . It computes the colour gradients in a random 500x297 pixel image using an intermediate table of about 1.5M rows with 9 element partitions, each sorted on two axes. It is also the test case that led me to work on this issue. Should I force parallelism and move it?

Mytherin replied:
Sure, please do.

hawkfish replied:
I've made a set of tests with about 1M rows that checks some boundary conditions around block sizes and copying strings across thread boundaries.  Coming shortly.

hawkfish replied:
Well that test appears to segfault under Linux 32-bit. Is it possible for someone to run it on that hardware? Win32 seems fine.

Mytherin replied:
The Linux 32-bit tests run on a 64-bit system, just with the FORCE_32_BIT flag enabled. You can check out the exact command [here](https://github.com/cwida/duckdb/blob/master/.github/workflows/main.yml#L353). When I ran into any failures on the CI there I always got the same failure running that script on a regular 64-bit Linux machine. Perhaps give that a shot (assuming you have access to a regular Linux box)?

hawkfish replied:
I don't but I'll see if I can make one of my Macs do it.  It's almost certainly something in the thread-safe copy code I missed.

hawkfish replied:
My Macs are all running Big Sur, which will not run 32 bit apps. Sigh.

Mytherin replied:
Perhaps try running with the thread sanitizer to see if that flags anything?

hawkfish replied:
Sadly no.  My Big Sur Mac won't link 32-bit app, but I have an old one at home I might be able to convince to reproduce it. A friend has an Ubuntu box I can borrow but not until next week.

hawkfish replied:
My old office iMac is too old to build with (10.10!) And the thread sanitiser didn't flag anything.

If anyone can have a quick look at this I'd appreciate it. I suspect that someone who has a Linux box and is familiar with the platform debuggers could find the problem pretty quickly. If not, I can probably muddle through some time next week.

hawkfish replied:
The issue shows up in debug builds and I think it is not caused by my CL.  if you look at the destructor for `WindowSegmentTree::~WindowSegmentTree`, it is doing pointer arithmetic assuming 64 bit pointers. When I look at the `addresses` array, the odd values are `0xffffffff`.

hawkfish replied:
That fixed it.

Mytherin reviewed (commented):
Thanks for the PR! Some comments from my side below.

re:the chunk collection issue; I think it is fine to keep it there for now. Having this API present already on the chunk collection will be helpful when we integrate the new sorting code in the window functions. The API clarifies the requirements for the sorting rework there and will make this integration easier.

hawkfish commented on src/execution/operator/aggregate/physical_window.cpp:
```diff
@@ -693,21 +881,80 @@ static void ComputeWindowExpression(BoundWindowExpression *wexpr, ChunkCollectio
+	}
+
+	void Finish() {
+		while (tasks_completed < task_count) {
```
What should it do while it is waiting? Just loop?  I assumed that the main thread had to do some work too...

hawkfish replied:
Yes, it just hangs.

hawkfish replied:
Ugh @hannesmuehleisen that doesn't fail for me.

hawkfish replied:
Ah, it fails in release mode.

hawkfish replied:
@hannesmuehleisen It looks like two different threads are trying to release the same `VectorBuffer` from the loop in `MaterializeExpressions` The threads are both reading the same block from the input, so is some buffer getting handed around under the assumption that only one thread will use it at a time? There is a destructor for a class called `single_thread_ptr` but I assumed that each `DataChunk` was being allocated independently on its thread and that the first argument of `ExpressionEvaluator::Evaluate` was effectively read-only/constant.  Is that not true?

hawkfish replied:
Ah, when there is a constant expression in the input then a reference is just shared between the input and the expression. Which is not thread safe.

hawkfish replied:
Same for referenced columns. It appears that the only way to share read-only column data between threads is to copy it.

Mytherin replied:
> 
> 
> What should it do while it is waiting? Just loop? I assumed that the main thread had to do some work too...

By default the main thread should be looping over tasks and executing them until they are all completed, hence just handing control back to that main loop should result in the same behavior (but without needing to duplicate the loop).

hawkfish replied:
At the bottom of the loop, I need to wait for all the function evaluation `Tasks to finish. How should that be implemented? A spin-loop seems like a waste of a thread - is there a semaphore system that could be leveraged?

hawkfish commented on src/execution/operator/aggregate/physical_window.cpp:
```diff
@@ -693,21 +881,80 @@ static void ComputeWindowExpression(BoundWindowExpression *wexpr, ChunkCollectio
 
+class TaskCounter {
+public:
+	explicit TaskCounter(TaskScheduler &scheduler_p)
```
That's what I was thinking, but I didn't want to presume without asking!

hannes committed:
checked in parallel window test from h2o benchmark

hannesmuehleisen commented:
Checked in the test case with a smaller version of the dataset into this branch: https://github.com/hannesmuehleisen/duckdb/tree/evilwindow

You could merge with this branch to get this test into this PR.

hawkfish commented:
After having a good look at everything, I think this CL is DOA.  I made a fundamental assumption that reading from a `Vector` was a `const` operation and this does not appear to be true. In addition, I was assuming that disjoint _writes_ to the output were thread safe too, which is not correct when strings are involved.

Instead, I think that multithreading `PhysicalWindow` will involve:

* Evaluating the partition and sort expressions on a single thread;
* Hashing the partition keys and computing the partition sizes (on a single thread?);
* Copying data out into partitions on separate threads (assuming that `VectorOperations::Copy` is actually thread-safe on the input side, which might be a separate change as in issue #1511);
* Computing _all_ functions for a single partition on the same thread;
* Concatenating the partitioned input and output data sequentially when complete

This will result in non-deterministic ordering for the output, which may make some test output non-deterministic. If the planner assumes that the window operator produces sorted output, this will need to change.

My big question for my reviewers is: Should I just throw (most of) this CL out and start over so the code review is easier?  A related question is: Should I take some of the fragments that will result (`TaskCounter`, the segment tree fix, `const` copying and hashing) and break them out into smaller commits?

Commit:
Merge branch 'hawkfish-window-partition' into hawkfish-const-copy

Commit:
Issue #1367: Window parallel partitions

Change the parallel execution model to
work on hash buckets. This was needed
because the previous model assumed that
the input chunks could be computed with
and sorted in parallel, which is not the case.

The new model computes the OVER fields
for each (OPK, OSK) pair, hashes the OPK fields and
then spins up Tasks to partition the data
for the non-empty buckets and compute the
window functions for each bucket. When all
Tasks have completed, the resulting collections
are Merged back into the operator state.

One important side-effect of this change is that
output ordering for tests with OVER clauses
is no longer deterministic. Such tests need to
use the rowsort specification to ensure that
the expected and actual output match.

Commit:
Issue #1367: Window parallel partitions

Force the slow SQL_Server tests to use parallelism.

Commit:
Issue #1367: Window hash partitions

Remove extra blank line.

hawkfish committed:
Merge pull request #35 from hannesmuehleisen/evilwindow

Issue #1367: Window hash partitions

Commit:
Issue #1367: Window hash partitions

Merge branch 'hawkfish-window-partition' of github.com:hawkfish/duckdb into hawkfish-window-partition

Commit:
Issue #1367: Window hash partition

Fix tidy issues.

Commit:
Issue #1367: Window hash partition

Fix misplaced ampersand.

Commit:
Issue #1367: Window hash partitions

Move slow test case with lots of
small partitions to parallel
test directory and force parallelism.

hawkfish committed:
Merge pull request #36 from cwida/master

Issue #1367: Window hash partition

Commit:
Issue #1367: Window hash partition

Add test cases for partitioning and string copying.

Commit:
Issue #1367: Window hash partition

Formatting.

hawkfish committed:
Merge pull request #38 from cwida/master

Issue #1367: Window hash partition

Commit:
Issue #1367: Window hash partition

Add order by to make test deteministic.

hawkfish committed:
Merge pull request #39 from cwida/master

Issue #1367: Window hash partition

Commit:
Issue #1367: Window hash partition

Fix pointer size problem in ~WindowSegmentTree.

hawkfish commented:
I think this is ready now.

Mytherin commented on test/sql/window/test_window_walmart.test_slow:
```diff
@@ -0,0 +1,63 @@
+statement ok
+PRAGMA threads=4
+
+statement ok
```
Should the result here be checked?

hawkfish replied:
I think the point was it crashed.

Mytherin replied:
Fair enough. I think this can be merged then. @hannesmuehleisen any more comments?

hannesmuehleisen replied:
If it passes the test i had added i'm happy

Mytherin reviewed (commented):
Awesome! Looks great. One minor comment:

hawkfish committed:
Merge pull request #44 from cwida/master

Issue #1367: Window hash partition
